datafusion/physical-optimizer/src/progressive_evaluation/statistics.rs

```diff
@@ -0,0 +1,294 @@
+    fn extract_statistics_from_file_group(
+        &self,
+        datasrc_exec: &DataSourceExec,
+    ) -> DatafusionResult<Option<ColumnStatistics>> {
+        let Ok(col_idx) = datasrc_exec.schema().index_of(self.column_name) else {
+            // maybe alias
+            return Ok(None);
+        };
+
+        // Extract partitioned files from the ParquetExec
+        let Some(base_config) = datasrc_exec
+            .data_source()
+            .as_any()
+            .downcast_ref::<FileScanConfig>()
+        else {
+            return Ok(None);
+        };
+
+        let file_group_partition = base_config.file_groups.get(self.partition).ok_or(
+            internal_datafusion_err!(
+                "requested partition does not exist in datasource exec"
+            ),
+        )?;
+
+        Ok(file_group_partition
+            .statistics()
+            .as_ref()
+            .and_then(|stats| stats.column_statistics.get(col_idx).cloned()))
+    }
+
+    /// Extract from datasource, with consideration of partitioning.
+    fn extract_from_data_source(
+        &mut self,
+        datasrc: &Arc<dyn ExecutionPlan>,
+    ) -> DatafusionResult<Option<ColumnStatistics>> {
+        if datasrc.as_any().downcast_ref::<EmptyExec>().is_some()
+            || datasrc
+                .as_any()
+                .downcast_ref::<PlaceholderRowExec>()
+                .is_some()
+        {
+            Ok(self.statistics.clone())
+        } else if let Some(datasrc_exec) =
+            datasrc.as_any().downcast_ref::<DataSourceExec>()
+        {
+            let datarsc_stats =
+                match self.extract_statistics_from_file_group(datasrc_exec) {
+                    Ok(Some(col_stats)) => Some(convert_min_max_to_exact(col_stats)),
+                    _ => None,
+                };
+            Ok(datarsc_stats)
+        } else {
+            // specific constraint on our plans
+            unreachable!("should not have another unsupported data source")
+        }
+    }
+}
+
+impl<'n> TreeNodeVisitor<'n> for PartitionStatisticsVisitor<'_> {
+    type Node = Arc<dyn ExecutionPlan>;
+
+    fn f_down(&mut self, node: &'n Self::Node) -> DatafusionResult<TreeNodeRecursion> {
+        if !is_supported(node) {
+            self.statistics = None;
+            Ok(TreeNodeRecursion::Stop)
+        } else if is_leaf(node) {
+            self.statistics = self.extract_from_data_source(node)?;
+            Ok(TreeNodeRecursion::Stop)
+        } else if should_merge_partition_statistics(node) {
+            self.statistics = self.merge_column_statistics_across_partitions(node)?;
+            Ok(TreeNodeRecursion::Jump)
+        } else if should_pass_thru_partition_statistics(node) {
+            self.statistics =
+                self.find_stats_per_partition_within_multiple_children(node)?;
+            Ok(TreeNodeRecursion::Stop)
+        } else {
+            self.statistics = None;
+            Ok(TreeNodeRecursion::Stop)
+        }
```

Comment:
This is where I tried to general the rules of how to extract partition min/max, without special casing each node type.

I prefer the approach suggested by xudong to have a caller on the `ExecutionPlan::statistics_per_partition`.

Reply:
I'm gonna merge the PR: https://github.com/apache/datafusion/pull/15852, then you can update the PR to use it.

Reply:
Thanks @xudong963 .

I ended up quickly porting over our own `statistics_per_partition` code, since it handles a few use cases we need. (Note: it doesn't handle joins and a few other use cases, which you have covered.)

I pushed it all up anyways, for the purposes of sharing code.